api/v1/standalone-dns-proxy/standalone-dns-proxy.proto

```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+syntax = "proto3";
+
+package standalonednsproxy;
+
+option go_package = "github.com/cilium/cilium/api/v1/standalonednsproxy";
+
+service FQDNData {
+  rpc SubscribeToDNSPolicies(stream  DNSPoliciesResult) returns (stream DNSPolicies) {}
```

Comment:
nit: inconsistent plural/singular between DNSPoliciesResult and DNSPolicies. Prefer singular, the `stream` already implies that it'll be "a stream of".

Since this is a bi-directional streaming request, it's important to document how the client- and server-originated messages correlate. The SDS (client) initiates the request, but who sends the first resource? Is the agent (server) allowed to send resources without having received an ACK for the previous?

Implementation-wise, making this a stream has certain implications: writing to a gRPC stream _must_ be serialized. What implications does this have for policy computation in the agent, what implications does it have for the SDS?

Reply:
>Since this is a bi-directional streaming request, it's important to document how the client- and server-originated messages correlate. The SDS (client) initiates the request, but who sends the first resource?

The messages are correlated via request id. The SDP(client) will call the `SubscribeToDNSPolicies` through the client interface and stores the stream returned from it. SDP will continue to listen(`recv`) the data from this stream.
Cilium Agent also stores the stream when SDP calls it the first and sends the first resource on the same stream. This way we are writing only from agent end and reading from it at the SDP end.

>Is the agent (server) allowed to send resources without having received an ACK for the previous?

The implementation of sending data from Cilium agent requires a read lock to send the snapshot of the policy rules until it is sent to SDP. And listens for ack in separate go routine for the ack.  So there can be data sent to SDP multiple time without receiving the ack, but since there is a lock mechanism in place the data is sent in a serialized manner. It is the responsibility of the SDP to make sure it reads the data from the stream. In case it fails, it will retry the above mentioned logic and receives the latest snapshot from cilium agent.

> Implementation-wise, making this a stream has certain implications: writing to a gRPC stream must be serialized. What implications does this have for policy computation in the agent, what implications does it have for the SDS?

The policy computation is done during endpoint regeneration and only after successful computation we are sending the data to SDP. With a read lock on the policyCache(which I think already happens) we avoid concurrent read of the current snapshot of the policy rules.
SDP in this case is dumb and reads the data from the stream and replaces the current snapshot of policies it has stored.